File: engine/src/main/java/com/terralite/engine/physics/GravitySystem.java
package com.terralite.engine.physics;

import com.terralite.engine.entity.Entity;
import com.terralite.engine.simulation.SimulationTick;
import com.terralite.engine.simulation.WorldSimulationSystem;
import com.terralite.engine.world.World;

public final class GravitySystem implements WorldSimulationSystem {
    private static final double GRAVITY = -20.0;
    private static final double TERMINAL_VELOCITY = -50.0;

    @Override
    public void tick(World world, SimulationTick tick) {
        double dt = tick.delta().toNanos() / 1_000_000_000.0;

        for (Entity entity : world.entities().entities()) {
            if (!entity.has(PhysicsComponents.VELOCITY)) continue;

            boolean grounded = entity.get(PhysicsComponents.GROUNDED).orElse(false);
            if (grounded) continue;

            Velocity v = entity.require(PhysicsComponents.VELOCITY);
            double newVy = Math.max(TERMINAL_VELOCITY, v.y() + GRAVITY * dt);
            entity.set(PhysicsComponents.VELOCITY, new Velocity(v.x(), newVy, v.z()));
        }
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/physics/GravitySystem.kt
package com.terralite.engine.physics

import com.terralite.engine.simulation.SimulationTick
import com.terralite.engine.simulation.WorldSimulationSystem
import com.terralite.engine.world.World

class GravitySystem : WorldSimulationSystem {
    override fun tick(world: World, tick: SimulationTick) {
        val dt = tick.delta / 1_000_000_000.0

        for (entity in world.entities().entities()) {
            if (!entity.has(PhysicsComponents.VELOCITY)) continue
            val grounded = entity.get(PhysicsComponents.GROUNDED) ?: false
            if (grounded) continue
            val v = entity.require(PhysicsComponents.VELOCITY)
            val newVy = maxOf(TERMINAL_VELOCITY, v.y + GRAVITY * dt)
            entity.set(PhysicsComponents.VELOCITY, Velocity(v.x, newVy, v.z))
        }
    }

    companion object {
        private const val GRAVITY = -20.0
        private const val TERMINAL_VELOCITY = -50.0
    }
}
